packages/knip/src/manifest/helpers.ts

```diff
@@ -1,12 +1,87 @@
 import type { Scripts } from '../types/package-json.js';
-import { join } from '../util/path.js';
+import { isFile } from '../util/fs.js';
+import { dirname, join } from '../util/path.js';
 import { _require } from '../util/require.js';
 
+const pnpStatus = {
+  dir: '',
+  pnpPath: '',
+  enabled: false,
+};
+
 type LoadPackageManifestOptions = { dir: string; packageName: string; cwd: string };
 
+const findNearestPnPFile = (startDir: string) => {
+  // Find the nearest .pnp.cjs file by traversing up
+  let currentDir = startDir;
+  while (currentDir !== '/') {
+    const pnpPath = join(currentDir, '.pnp.cjs');
+    if (isFile(pnpPath)) {
+      const pnpApi = _require(pnpPath);
+      pnpApi.setup();
+      pnpStatus.dir = startDir;
+      pnpStatus.pnpPath = pnpPath;
+      pnpStatus.enabled = true;
+      return;
+    }
+    // Move up one directory
+    const parentDir = dirname(currentDir);
+    if (parentDir === currentDir) {
+      break; // Reached root
+    }
+    currentDir = parentDir;
+  }
+  pnpStatus.dir = startDir;
+  pnpStatus.pnpPath = '';
+  pnpStatus.enabled = false;
+};
```

Comment:
I made this by assuming that knip needs to be able to understand yarn pnp, even it isn't under yarn pnp environment, e.g `yarn dlx knip`. If it's not, we can just simply import pnpapi directly, instead of manually finding it.